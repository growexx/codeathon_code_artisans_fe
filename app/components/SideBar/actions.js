import {
  ADD_SIDEBAR_ITEM,
  LOAD_SIDEBAR_ITEMS,
  LOAD_SIDEBAR_ITEMS_SUCCESS,
  REMOVE_SIDEBAR_ITEM,
} from './constants';

export const addSidebarItem = data => ({
  type: ADD_SIDEBAR_ITEM,
  data,
});

export const removeSidebarItem = id => ({
  type: REMOVE_SIDEBAR_ITEM,
  id,
});

export const loadSidebarItems = () => ({
  type: LOAD_SIDEBAR_ITEMS,
});

export const loadDataSuccess = data => ({
  type: LOAD_SIDEBAR_ITEMS_SUCCESS,
  data,
});
