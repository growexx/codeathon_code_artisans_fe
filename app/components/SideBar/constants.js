import React from 'react';
import { MessageOutlined } from '@ant-design/icons';
import {
  ROUTES,
  ROLE_BASED_SIDEBAR_MENU,
  ROLES,
  RESTRICTED_ROUTES,
} from 'containers/constants';
import { Skeleton } from 'antd';
import { isArray } from 'lodash';

export const ADD_SIDEBAR_ITEM = 'llm/Sidebar/ADD_SIDEBAR_ITEM';
export const REMOVE_SIDEBAR_ITEM = 'llm/Sidebar/REMOVE_SIDEBAR_ITEM';
export const LOAD_SIDEBAR_ITEMS = 'llm/Sidebar/LOAD_SIDEBAR_ITEMS';
export const LOAD_SIDEBAR_ITEMS_SUCCESS =
  'llm/Sidebar/LOAD_SIDEBAR_ITEMS_SUCCESS';

export const GET_FILTERED_MENU_ITEM = (role, data) => {
  let menuItems;
  if (isArray(data)) {
    menuItems = data?.map(item => ({
      id: item.chatId,
      to: `/chat/${item.chatId}`,
      tabName: item.name,
      icon: <MessageOutlined />,
    }));
  } else {
    menuItems = ['a', 'b', 'c', 'd'];
    menuItems = menuItems.map((item, index) => ({
      id: item,
      to: '',
      tabName: <Skeleton active title paragraph={false} />,
      icon: null,
    }));
  }
  return menuItems;
};
export const GA_LABEL_SIDEBAR = 'SideBar';
