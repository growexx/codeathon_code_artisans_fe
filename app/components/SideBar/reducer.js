/* eslint-disable consistent-return */
/* eslint-disable no-param-reassign */
import produce from 'immer';
import {
  ADD_SIDEBAR_ITEM,
  LOAD_SIDEBAR_ITEMS_SUCCESS,
  REMOVE_SIDEBAR_ITEM,
} from './constants';

export const initialState = {
  menuItems: [],
};

const sidebarReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case ADD_SIDEBAR_ITEM:
        draft.menuItems.unshift(action.data);
        break;
      case REMOVE_SIDEBAR_ITEM:
        draft.menuItems = draft.menuItems.filter(
          item => item.chatId !== action.id,
        );
        break;
      case LOAD_SIDEBAR_ITEMS_SUCCESS:
        draft.menuItems = action.data;
        break;
      default:
        return state;
    }
  });

export default sidebarReducer;
