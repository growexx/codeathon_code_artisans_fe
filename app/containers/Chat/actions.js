import {
  GET_CHAT_HISTORY,
  ADD_CHAT_QUESTION,
  SET_CHAT_HISTORY,
  ADD_CHAT_ANSWER,
} from './constants';

export const getChatHistory = () => ({
  type: GET_CHAT_HISTORY,
});

export const addChatQuestion = (chatId, question) => ({
  type: ADD_CHAT_QUESTION,
  chatId,
  question,
});

export const addChatAnswer = (chatId, answer, sources) => ({
  type: ADD_CHAT_ANSWER,
  chatId,
  answer,
  sources,
});

export const setChatHistory = (chatId, history) => ({
  type: SET_CHAT_HISTORY,
  chatId,
  history,
});
