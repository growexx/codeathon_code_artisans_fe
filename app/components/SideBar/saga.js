import { call, put, takeLatest } from 'redux-saga/effects';
import request from 'utils/request';
import { LOAD_SIDEBAR_ITEMS } from './constants';
import { loadDataSuccess } from './actions';
import { API_ENDPOINTS } from '../../containers/constants';

export function* fetchSidebarItems() {
  const response = yield call(request, API_ENDPOINTS.CHAT_INFO, {
    method: 'GET',
  });
  if (response.status === 1) {
    yield put(loadDataSuccess(response.data));
  }
}

export default function* watchLoadSidebar() {
  yield takeLatest(LOAD_SIDEBAR_ITEMS, fetchSidebarItems);
}
