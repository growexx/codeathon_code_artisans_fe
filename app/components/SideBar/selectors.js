import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectData = state => state.sidebar || initialState;

const selectMenuItems = () =>
  createSelector(
    selectData,
    dataState => dataState.menuItems,
  );

const selectError = () =>
  createSelector(
    selectData,
    dataState => dataState.error,
  );

export { selectMenuItems, selectError };
