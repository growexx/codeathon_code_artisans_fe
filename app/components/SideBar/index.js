/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { useEffect, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Button, Layout, Menu, Modal } from 'antd';
import {
  PlusOutlined,
  DeleteOutlined,
  MenuFoldOutlined,
} from '@ant-design/icons';
import PropTypes from 'prop-types';
import { v4 as uuid } from 'uuid';
import request from 'utils/request';
import { useInjectReducer } from 'utils/injectReducer';
import { useInjectSaga } from 'utils/injectSaga';
import { GA_LABEL_SIDEBAR, GET_FILTERED_MENU_ITEM } from './constants';
import { eventGA } from '../../utils/googleAnalytics';
import { GA_CATEGORY_MENU_CLICKS } from '../../utils/constants';
import { selectError, selectMenuItems } from './selectors';
import reducer from './reducer';
import saga from './saga';
import { loadSidebarItems, removeSidebarItem } from './actions';
import { API_ENDPOINTS } from '../../containers/constants';

const { Sider } = Layout;
const key = 'sidebar';

const SideBar = ({
  collapsed,
  user,
  menuItems,
  getSidebarItems,
  removeItem,
  toggle,
}) => {
  useInjectReducer({ key, reducer });
  useInjectSaga({ key, saga });

  const location = useLocation();
  const navigate = useNavigate();
  const uniqueId = uuid().slice(0, 6);
  const [modalVisible, setModalVisible] = useState(false);
  const [loading, setLoading] = useState(false);
  const [selectedOption, setSelectedOption] = useState(
    location.pathname.split('/')[2],
  );

  const handleOptionClick = option => {
    setSelectedOption(option);
    navigate(`/chat/${option}`);
  };

  const handleDeleteClick = () => {
    setModalVisible(true);
  };

  const handleModalConfirm = async () => {
    setLoading(true);
    await request(`${API_ENDPOINTS.DELETE_CHAT}?id=${selectedOption}`, {
      method: 'GET',
    });
    removeItem(selectedOption);
    setModalVisible(false);
    setLoading(false);
    navigate(`/new-chat/${uniqueId}`);
  };

  const handleModalCancel = () => {
    setModalVisible(false);
  };

  const handleNewChatClick = () => {
    navigate(`/new-chat/${uniqueId}`);
  };

  useEffect(() => {
    getSidebarItems();
  }, []);

  useEffect(() => {
    setSelectedOption(location.pathname.split('/')[2]);
  }, [location.pathname.split('/')[2]]);

  return (
    <>
      <div className={!collapsed ? 'extra-content' : ''} />
      <Sider
        trigger={null}
        collapsible
        collapsed={collapsed}
        collapsedWidth={0}
        width={244}
        id="components-layout-demo-custom-trigger"
        className={!collapsed ? 'not-collapsed-sider' : ''}
      >
        <div className="new-chat-wrapper">
          <Button
            type="primary"
            className="new-chat-btn"
            icon={<PlusOutlined />}
            onClick={handleNewChatClick}
          >
            New Chat
          </Button>
          <Button
            type="primary"
            className="sidebar-btn"
            data-testid="SIDEBAR_TOGGLE"
            icon={<MenuFoldOutlined />}
            onClick={toggle}
          />
        </div>
        <Menu
          theme="dark"
          mode="inline"
          defaultSelectedKeys={[location.pathname.split('/')[2]]}
          selectedKeys={[location.pathname.split('/')[2]]}
        >
          {GET_FILTERED_MENU_ITEM(user && user.role, menuItems).map(menu => (
            <Menu.Item
              key={menu.id}
              icon={menu.icon}
              onClick={() => {
                eventGA(
                  GA_CATEGORY_MENU_CLICKS,
                  `${menu.to} clicked from sidebar`,
                  GA_LABEL_SIDEBAR,
                );
                handleOptionClick(menu.id);
              }}
            >
              <span onClick={() => handleOptionClick(menu.id)}>
                {menu.tabName}
              </span>
              {selectedOption === menu?.id.toString() && (
                <DeleteOutlined
                  onClick={handleDeleteClick}
                  data-testid="DELETE_SIDEBAR"
                  style={{
                    position: 'absolute',
                    right: '10px',
                    top: '13px',
                  }}
                />
              )}
            </Menu.Item>
          ))}
        </Menu>
        <Modal
          centered
          title="Delete chat?"
          okText="Delete"
          visible={modalVisible}
          onOk={handleModalConfirm}
          onCancel={handleModalCancel}
          closeIcon={false}
          width={448}
          confirmLoading={loading}
        >
          Are you sure you want to delete this chat history?
        </Modal>
      </Sider>
    </>
  );
};

SideBar.propTypes = {
  collapsed: PropTypes.bool,
  user: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
  menuItems: PropTypes.any,
  getSidebarItems: PropTypes.func,
  removeItem: PropTypes.func,
  toggle: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  menuItems: selectMenuItems(),
  error: selectError(),
});

export function mapDispatchToProps(dispatch) {
  return {
    getSidebarItems: () => dispatch(loadSidebarItems()),
    removeItem: id => dispatch(removeSidebarItem(id)),
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(withConnect)(SideBar);
