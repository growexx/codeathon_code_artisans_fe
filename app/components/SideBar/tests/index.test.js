import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { HistoryRouter as Router } from 'redux-first-history/rr6';
import history from 'utils/history';
import request from 'utils/request';
import SideBar, { mapDispatchToProps } from '../index';
import configureStore from '../../../configureStore';
import { loadDataSuccess } from '../actions';
import 'jest-dom/extend-expect';

jest.mock('utils/request');
let globalStore;
const props = {
  user: {
    role: 1,
  },
  collapsed: true,
};
const componentWrapper = () =>
  render(
    <Provider store={globalStore}>
      <IntlProvider locale="en">
        <Router history={history}>
          <SideBar {...props} />
        </Router>
      </IntlProvider>
    </Provider>,
  );

describe('<SideBar />', () => {
  beforeAll(() => {
    const { store } = configureStore({});
    globalStore = store;
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = componentWrapper();
    expect(firstChild).toMatchSnapshot();
  });
  it('mapDispatch to props', () => {
    const mockFn = jest.fn();
    const eventObject = {
      target: {
        value: 'test',
      },
      preventDefault: jest.fn(),
    };
    const returnValue = mapDispatchToProps(mockFn);
    returnValue.getSidebarItems(eventObject);
    returnValue.removeItem(eventObject);
    expect(mockFn).toBeCalled();
  });
  it('delete sidebar item', async () => {
    request.mockResolvedValueOnce({ status: 1 });
    globalStore.dispatch(
      loadDataSuccess([
        {
          chatId: '2e75',
          name: 'Question',
        },
      ]),
    );
    props.collapsed = false;
    const { getByText, getByTestId, queryByText } = componentWrapper();
    await waitFor(() => {
      expect(getByText('Question')).toBeInTheDocument();
    });
    fireEvent.click(getByText('Question'));
    await waitFor(() => {
      fireEvent.click(getByTestId('DELETE_SIDEBAR'));
      fireEvent.click(getByText('Delete'));
    });
    expect(queryByText('Question')).not.toBeInTheDocument();
  });
  it('donot delete sidebar item', async () => {
    request.mockResolvedValueOnce({ status: 1 });
    globalStore.dispatch(
      loadDataSuccess([
        {
          chatId: '2e75',
          name: 'Question',
        },
      ]),
    );
    const { getByText, getByTestId, queryByText } = componentWrapper();
    await waitFor(() => {
      expect(getByText('Question')).toBeInTheDocument();
    });
    fireEvent.click(getByText('Question'));
    await waitFor(() => {
      fireEvent.click(getByTestId('DELETE_SIDEBAR'));
      fireEvent.click(getByText('Cancel'));
    });
    expect(queryByText('Question')).toBeInTheDocument();
  });
  it('new chat', async () => {
    request.mockResolvedValueOnce({ status: 1 });
    globalStore.dispatch(
      loadDataSuccess([
        {
          chatId: '2e75',
          name: 'Question',
        },
      ]),
    );
    const { getByText, getByTestId, queryByText } = componentWrapper();
    await waitFor(() => {
      expect(getByText('Question')).toBeInTheDocument();
    });
    fireEvent.click(getByTestId('SIDEBAR_TOGGLE'));
    fireEvent.click(getByText('New Chat'));
    expect(queryByText('Question')).toBeInTheDocument();
  });
});
