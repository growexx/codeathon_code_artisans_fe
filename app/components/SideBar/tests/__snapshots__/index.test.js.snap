// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideBar /> should render and match the snapshot 1`] = `
<div
  class=""
/>
`;
